File: agents/go/internal/workflow/workflow.go
// Package workflow builds a deterministic triage → diagnose → recommend pipeline (Chapter 3.5).
//
// Where the root agent (an llmagent) decides its own steps, a SequentialAgent runs a fixed
// pipeline: each sub-agent runs in order, passing its findings forward through session state.
// In Go this is the idiomatic workflow-agents API (sequentialagent / parallelagent / loopagent);
// note that Python's classic workflow agents are deprecated in favor of its graph Workflow —
// the same pipeline, different SDK surface.
package workflow

import (
	"context"
	"fmt"

	"google.golang.org/genai"

	adkagent "google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/agent/llmagent"
	"google.golang.org/adk/v2/agent/workflowagents/sequentialagent"
	"google.golang.org/adk/v2/model/gemini"
	"google.golang.org/adk/v2/tool"

	"github.com/MLOps-Courses/agentops-open-course/agents/go/internal/config"
	"github.com/MLOps-Courses/agentops-open-course/agents/go/internal/data"
	"github.com/MLOps-Courses/agentops-open-course/agents/go/internal/memory"
	"github.com/MLOps-Courses/agentops-open-course/agents/go/internal/tools"
)

const (
	triageInstruction = "You triage incidents. List the unresolved incidents and pick the single most " +
		"urgent one (lowest SEV number wins). State its id, service, severity, and one-line summary."
	diagnoseInstruction = "You diagnose the incident chosen by triage. Use get_incident for its details " +
		"and its runbook (get_runbook), and get_service_status for the service. Explain the likely cause " +
		"in two or three sentences, citing the runbook."
	recommendInstruction = "You recommend remediation for the diagnosed incident. Using the runbook, give " +
		"a short, ordered list of next steps. Flag any step that needs a guarded action (restart_service, " +
		"resolve_incident) and requires human approval. Cite the runbook you used."
)

// New builds the triage → diagnose → recommend workflow over the shared dataset.
func New(ctx context.Context, cfg config.Config, store *data.Store) (adkagent.Agent, error) {
	model, err := gemini.NewModel(ctx, cfg.Model, &genai.ClientConfig{APIKey: cfg.APIKey})
	if err != nil {
		return nil, fmt.Errorf("creating gemini model %q: %w", cfg.Model, err)
	}

	readTools, err := tools.All(store)
	if err != nil {
		return nil, fmt.Errorf("building tools: %w", err)
	}
	knowledge, err := memory.KnowledgeTools(store)
	if err != nil {
		return nil, fmt.Errorf("building knowledge tools: %w", err)
	}
	diagnoseTools := make([]tool.Tool, 0, len(readTools)+len(knowledge))
	diagnoseTools = append(diagnoseTools, readTools...)
	diagnoseTools = append(diagnoseTools, knowledge...)

	triage, err := llmagent.New(llmagent.Config{
		Name: "triage", Model: model,
		Description: "Finds the most urgent unresolved incident.",
		Instruction: triageInstruction, Tools: readTools,
	})
	if err != nil {
		return nil, fmt.Errorf("building triage step: %w", err)
	}
	diagnose, err := llmagent.New(llmagent.Config{
		Name: "diagnose", Model: model,
		Description: "Explains the likely cause of the triaged incident.",
		Instruction: diagnoseInstruction, Tools: diagnoseTools,
	})
	if err != nil {
		return nil, fmt.Errorf("building diagnose step: %w", err)
	}
	recommend, err := llmagent.New(llmagent.Config{
		Name: "recommend", Model: model,
		Description: "Recommends concrete, runbook-backed remediation.",
		Instruction: recommendInstruction, Tools: knowledge,
	})
	if err != nil {
		return nil, fmt.Errorf("building recommend step: %w", err)
	}

	return sequentialagent.New(sequentialagent.Config{
		AgentConfig: adkagent.Config{
			Name:        "triage_workflow",
			Description: "Runs triage → diagnose → recommend over the current incidents.",
			SubAgents:   []adkagent.Agent{triage, diagnose, recommend},
		},
	})
}

File: agents/python/src/agent/workflow.py
"""A deterministic triage → diagnose → recommend workflow (Chapter 3.5).

Where ``root_agent`` (an LlmAgent) decides its own steps, a ``Workflow`` runs a fixed graph:
agents are nodes, and an edge chains them so each runs in order, passing findings forward via
session state. ``Workflow`` is ADK 2.0's graph runtime — it supersedes the classic
``SequentialAgent`` / ``ParallelAgent`` / ``LoopAgent`` (now deprecated) and also expresses
parallel, looping, and dynamic DAGs.
"""

from __future__ import annotations

from google.adk import Agent, Workflow
from google.adk.agents.llm_agent import ToolUnion

from .guardrails import handle_model_error, handle_tool_error, secure_tool_output
from .memory import KNOWLEDGE_TOOLS
from .model import build_model
from .pii import redact_request_pii, redact_response_pii
from .tools import ALL_TOOLS

# Each step is a focused agent; the Workflow chains them. Tools reflect what each step needs.
_DIAGNOSE_TOOLS: list[ToolUnion] = [*ALL_TOOLS, *KNOWLEDGE_TOOLS]

# 1) Triage: find the incident that matters most right now.
triage = Agent(
    model=build_model(),
    name="triage",
    description="Finds the most urgent unresolved incident.",
    instruction=(
        "You triage incidents. List the unresolved incidents and pick the single most urgent one "
        "(lowest SEV number wins). State its id, service, severity, and one-line summary."
    ),
    tools=ALL_TOOLS,
    before_model_callback=redact_request_pii,
    after_model_callback=redact_response_pii,
    after_tool_callback=secure_tool_output,
    on_model_error_callback=handle_model_error,
    on_tool_error_callback=handle_tool_error,
)

# 2) Diagnose: explain the likely cause of the triaged incident.
diagnose = Agent(
    model=build_model(),
    name="diagnose",
    description="Explains the likely cause of the triaged incident.",
    instruction=(
        "You diagnose the incident chosen by triage. Use get_incident for its details and its "
        "runbook (get_runbook), and get_service_status for the service. Explain the likely cause "
        "in two or three sentences, citing the runbook."
    ),
    tools=_DIAGNOSE_TOOLS,
    before_model_callback=redact_request_pii,
    after_model_callback=redact_response_pii,
    after_tool_callback=secure_tool_output,
    on_model_error_callback=handle_model_error,
    on_tool_error_callback=handle_tool_error,
)

# 3) Recommend: propose concrete, runbook-backed next steps.
recommend = Agent(
    model=build_model(),
    name="recommend",
    description="Recommends concrete, runbook-backed remediation.",
    instruction=(
        "You recommend remediation for the diagnosed incident. Using the runbook, give a short, "
        "ordered list of next steps. Flag any step that needs a guarded action (restart_service, "
        "resolve_incident) and requires human approval. Cite the runbook you used."
    ),
    tools=KNOWLEDGE_TOOLS,
    before_model_callback=redact_request_pii,
    after_model_callback=redact_response_pii,
    after_tool_callback=secure_tool_output,
    on_model_error_callback=handle_model_error,
    on_tool_error_callback=handle_tool_error,
)

# The graph: START → triage → diagnose → recommend, sharing session state along the way.
triage_workflow = Workflow(
    name="triage_workflow",
    description="Runs triage → diagnose → recommend over the current incidents.",
    edges=[("START", triage, diagnose, recommend)],
)
